feature: realtime agents

nicko-ai opened the pull request:
- Add realtime orchestration entry points so agencies can run interactive sessions.
- Package fully working demos: hosted web UI stack, Twilio bridge, and interactive example under examples/interactive/realtime/.
- Document the new capability in docs/additional-features/, update site metadata/images, and surface the flow from the examples README.
- Cover the behavior with tests/integration/realtime/test_realtime.py to lock in the realtime pipeline.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Introduce realtime voice agents with a FastAPI websocket bridge, voice-aware agents, packaged web/Twilio demos, docs, and tests.
> 
> - **Core runtime (realtime)**:
>   - Add `RealtimeAgency`/`RealtimeAgent` wrappers and `Agency.to_realtime()` for OpenAI Realtime sessions.
>   - New `integrations.run_realtime` FastAPI bridge with `/realtime` websocket, voice switching, turn detection, audio formats, Twilio media-stream support.
>   - Agent voice support (`voice` param, allowed voices) and agency-level voice randomization (`randomize_agent_voices`, `voice_random_seed`).
> - **FastAPI**:
>   - `run_fastapi` can mount `/your_agency/realtime` websockets via `enable_realtime` + `realtime_options`.
> - **UI/Demos**:
>   - Packaged browser realtime demo (FastAPI app + WebAudio client) and `examples/interactive/realtime/demo.py` launcher.
>   - Twilio phone bridge demo with media stream handler and README.
> - **Docs**:
>   - New Voice Agents docs (Overview, Deployment); update FastAPI integration and API reference; docs navigation updated.
> - **Exports**:
>   - Re-export `run_realtime` and `realtime` modules from package.
> - **Tests**:
>   - Integration tests for realtime orchestration and FastAPI `/realtime` endpoint; agent/agency voice features.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit d5ee450af577b4a15b1f81a6995982efcc436327. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

bonk1t commented:
do not merge

ArtemShatokhin reviewed (approved):
Works well, the only minor issue I could find is that when sending images, the text response is not updated in real time and after a bit of delay, it gets printed in full right away.

bonk1t commented on src/agency_swarm/integrations/realtime.py:
```diff
@@ -0,0 +1,484 @@
+    raw_message: dict[str, Any] = {"type": msg_type}
+    if other:
+        raw_message["other_data"] = other
+    client_message = cast(RealtimeModelRawClientMessage, raw_message)
```
Bug report incorrect: the current implementation preserves the required top-level event schema.

- `src/agency_swarm/integrations/realtime.py:177` collects every non-type field into `other_data` before wrapping the payload in `RealtimeModelSendRawMessage`; nothing is dropped — it’s just temporarily grouped.  
- The SDK type explicitly documents this pattern: `agents/realtime/model_inputs.py:12` notes that `other_data` is “merged into the message body,” indicating that downstream helpers will re-flatten it.  
- That re-flattening occurs in `_ConversionHelper.try_convert_raw_message`, which rebuilds the original `{ "type": ..., ... }` structure before serializing to the wire (`agents/realtime/openai_realtime.py:959`).  
- Since the OpenAI client event validator runs on the flattened dictionary, any unsupported schema would fail validation there rather than being silently ignored.

VRSEN commented on docs/additional-features/realtime-agents.mdx:
```diff
@@ -0,0 +1,164 @@
+
+## How it works (at a glance)
+
+Client (mic/WebRTC or Twilio) ↔ **FastAPI WebSocket** ↔ **OpenAI Realtime** ↔ **Agency Swarm Agent + Tools**.
```
Nonono. Move everything related to agent deployment (fastapi websocker) into another page. You already lost even me on this line here. Only show how to create agents here and link deployment page at the bottom.

Please keeps docs as simple as possible even for complete beginners. Docs is the product

VRSEN commented on docs/additional-features/realtime-agents.mdx:
```diff
@@ -0,0 +1,164 @@
+
+## Add a simple web client (browser)
+
+Use the official OpenAI Realtime example client (WebRTC or WebSocket). Point it to `ws://localhost:8000/realtime` (or your deployed address), and you’ll get mic capture, response audio, and a minimal UI.
```
What client? I have no idea what this is even about

VRSEN commented on docs/additional-features/realtime-agents.mdx:
```diff
@@ -0,0 +1,164 @@
+
+See OpenAI’s example for a copy-paste front end and event payloads.
+
+## Twilio (phone calls) — optional
```
Also should be in deployment section

VRSEN commented on docs/additional-features/realtime-agents.mdx:
```diff
@@ -0,0 +1,164 @@
+    agency = Agency(voice_agent)
+    run_realtime(
+        agency=agency.to_realtime(),
+        model="gpt-realtime",
```
I like this part that I define the model here for all agents

VRSEN commented on docs/additional-features/realtime-agents.mdx:
```diff
@@ -0,0 +1,164 @@
+
+## Operational tips
+
+- Interrupts: With server VAD, the API emits interruption events; keep your UI/telephony bridge ready to truncate playback and resume.
```
I don't understand this at all, maybe show what interrupt even looks like

nicko-ai committed:
feat: add realtime voice agents

- add realtime session bridge with per-agent voice support
- mount /{agency}/realtime websocket routes via run_fastapi
- ship packaged browser + Twilio realtime demos
- document voice agents and FastAPI integration options
- add regression tests for voice config and realtime endpoints

nicko-ai committed:
fix: align realtime typing and integration tests

nicko-ai committed:
feat: make realtime demo provider-aware for voice QA

nicko-ai committed:
fix: include xai realtime model in websocket url

nicko-ai committed:
test: cover stale xai realtime model query

nicko-ai committed:
test: reset handoff deprecation warning flag
